### PR TITLE
2230-datapages-add-static-images-in-faq-section

### PR DESCRIPTION
**Pull request by mlbrgl**
- [x] is the branch up-to-date with master?
- [x] 
### <samp>🤖 Inspired by Copilot at cd6115e</samp>
This pull request adds support for static images to data pages content coming from Google Docs . It renames the `DataPageGdoc` type to `DataPageGdocContent` and imports the `parseGdocContentFromAllowedLevelOneHeadings` function to `baker/GrapherBaker.tsx`. It also passes the parsed content and the previewing flag as props to the `DataPage` and `GrapherPage` components.
- [x] what problem is being solved here?
Static images in data pages gdoc content are not rendered
- [x] what solution has been chosen?
Re-use the existing `context` infrastructure used on `OwidGdoc` to pass `imageMetadata` down to the `Image` component. 
- [x] are there other solutions?
- [x] is the scope clear, well-defined and small?
- [x] could a flow chart / sequence diagram / architecture diagram help?
- [x] what is the happy path?
  - add an image block to one of the sections of the FAQ of a data page ([gdoc edit](https://docs.google.com/document/d/1XHpp9OU-UHZWD515v3V1KDY-YlSU5_-Hs_4_wGCGm1k/edit#heading=h.tvgnagh9k5u1))
  - check image is showing in preview ([example](http://staging-site-2204-datapages-add-collapse-functionality-to-faq-s/admin/grapher/banana-production))
  - publish data page companion gdoc ([gdoc preview](http://staging-site-2230-datapages-add-static-images-in-faq-section/admin/gdocs/1XHpp9OU-UHZWD515v3V1KDY-YlSU5_-Hs_4_wGCGm1k/preview)). Make sure to publish as a fragment by adding `type:fragment` in the gdoc frontmatter.
  - Mark the [datapage JSON](https://github.com/owid/owid-content/blob/master/datapages/587683.json) as ready to publish (`"status":"published"`). ⚠️ Only do this locally, do not commit.
  - check image is showing when baking ([example](http://staging-site-2204-datapages-add-collapse-functionality-to-faq-s/grapher/banana-production))
- [x] could the code be easier to read and understand?
- [x] can I help increase our confidence in the implementation by contributing tests?
- [x] should some errors be made more prominent?
- [x] what is missing, what should have been added / done / changed but hasn't?
  - [x] test baking

### CODE REVIEW

**Comment by mlbrgl** [comment]
Follow-ups:
- #2243 
- @ikesau should we validate published fragments in `bakeGdocPosts()`? Right now, they are neither baked (which is good) nor validated (which is bad).

**Comment by mlbrgl** [comment]
for a faster localBake, patch to only bake the banana-production var and chart: 
[bake-banana-production-var-and-chart-only.patch](https://github.com/owid/owid-grapher/files/11545325/bake-banana-production-var-and-chart-only.patch)

**Comment by mlbrgl** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2233** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2233" target="_blank">Graphite</a>  👈
    * **PR #2248** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2248" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/2233?utm_source=stack-comment).

**Comment by ikesau** [comment]
Uhh is this diff correct now? Seems like there's lot of extra changes and your reply to my earlier question about using Gdrive has been gobbled up somehow.

File: baker/GrapherBaker.tsx [hunk 1]
```diff
@@ -109,28 +102,146 @@ export const renderDataPageOrGrapherPage = async (
...
-        isPreviewing,
+        true,
         publishedExplorersBySlug
     )
 
+    const datapageGdocContent =
+        parseGdocContentFromAllowedLevelOneHeadings(datapageGdoc)
+
+    // Passing `datapageGdocContent` as a separate prop rather than weaving it
+    // into datapageGdoc to keep the concept of "Gdoc" clearer
```
Comment (ikesau) [inline baker/GrapherBaker.tsx:234]: Just confirming, fragments that are using this heading technique are still having their links tracked, right?

File: baker/GrapherBaker.tsx [hunk 2]
```diff
@@ -109,28 +102,146 @@ export const renderDataPageOrGrapherPage = async (
...
     return renderToHtmlPage(
         <DataPage
             grapher={grapher}
             variableId={id}
             datapageJson={datapageJson}
             datapageGdoc={datapageGdoc}
+            datapageGdocContent={datapageGdocContent}
```
Comment (ikesau) [inline baker/GrapherBaker.tsx:241]: So previews won't have images?

**Review by ikesau — COMMENTED** [review]
Just a few little queries 🙂 

Also, what motivated splitting out the preview/baked versions like this? Of the discrepancies between the two functions that I could spot, I didn't see anything that made me go "Oh, this is why they can't be the same function"

File: datapage/Datapage.ts [hunk 3]
```diff
@@ -71,12 +71,16 @@ export const getDatapageJson = async (
...
+ *
+ * see https://github.com/owid/owid-grapher/issues/2121#issue-1676097164
  */
 export const getDatapageGdoc = async (
     datapageJson: DataPageJson,
     isPreviewing: boolean,
     publishedExplorersBySlug?: Record<string, ExplorerProgram>
-): Promise<DataPageGdoc | null> => {
+): Promise<OwidGdocInterface | null> => {
+    if (!datapageJson.googleDocEditLink) return null
```
[the remark this answers is not in the document]
Comment (mlbrgl) [reply datapage/Datapage.ts:82]: makes sense, thanks!

File: datapage/Datapage.ts [hunk 4]
```diff
@@ -100,13 +106,9 @@ export const getDatapageGdoc = async (
...
 
-    if (!rawDatapageGdoc) return null
+    if (!datapageGdoc) return null
 
-    // Split the gdoc content into fields using the heading one texts as field
-    // names. Any top level key that is not in AllowedDataPageGdocFields will
-    // be silently filtered out. Beyond that, any allowed field can contain the
-    // whole range of OwidEnrichedGdocBlock types.
-    return parseGdocContentFromAllowedLevelOneHeadings(rawDatapageGdoc)
+    return datapageGdoc
```
[the remark this answers is not in the document]
Comment (mlbrgl) [reply datapage/Datapage.ts]: good catch!

File: baker/GrapherBaker.tsx (hunk 1 above)
Comment (mlbrgl) [reply baker/GrapherBaker.tsx:246]: You're right to flag this, they don't at the moment, but [they should](https://github.com/owid/owid-grapher/pull/2233#issuecomment-1559173499).

File: baker/GrapherBaker.tsx (hunk 2 above)
Comment (mlbrgl) [reply baker/GrapherBaker.tsx:253]: [Previews call `Gdoc.getGdocFromContentSource`](https://github.com/owid/owid-grapher/blob/ba149c5a04355d634b1fe7c0947bbc8816e43529/datapage/Datapage.ts#L100) which loads `imageMetadata` on the gdoc instance, so they do show images. Or were you referring to something else?

**Comment by mlbrgl** [comment]
@ikesau 

> Also, what motivated splitting out the preview/baked versions like this?

The branching was getting a big too complex for my taste to reason with. Making sure a change in one scenario wouldn't impact the other wasn't straightforward enough. I'm not a big fan of the repetition either but I feel like in this case this is the lesser of two evils.

**Review by ikesau — APPROVED** [review]

**Comment by mlbrgl** [comment]
@mlbrgl started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/2233).

**Comment by mlbrgl** [comment]
@mlbrgl merged this pull request with [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/2233).

**Pull request merged**